cache/clock_cache.h

```diff
@@ -232,17 +371,129 @@ struct ClockHandle {
+
+  // Tries to take an external ref. Returns true iff it succeeds.
+  inline bool TryExternalRef() {
+    if (!((refs += kOneExternalRef) & (EXCLUSIVE_REF | WILL_BE_DELETED))) {
+      return true;
+    }
+    refs -= kOneExternalRef;
+    return false;
+  }
+
+  // Releases an external ref. Returns the new value (this is useful to
+  // avoid an extra atomic read).
+  inline uint32_t ReleaseExternalRef() { return refs -= kOneExternalRef; }
+
+  // Take an external ref, assuming there is already one external ref
+  // to the handle.
+  void Ref() {
+    // TODO(Guido) Is it okay to assume that the existing external reference
+    // survives until this function returns?
+    refs += kOneExternalRef;
+  }
+
+  // Tries to take an internal ref. Returns true iff it succeeds.
+  inline bool TryInternalRef() {
+    if (!((refs += kOneInternalRef) & (EXCLUSIVE_REF | WILL_BE_DELETED))) {
+      return true;
+    }
+    refs -= kOneInternalRef;
+    return false;
+  }
+
+  inline void ReleaseInternalRef() { refs -= kOneInternalRef; }
+
+  // Tries to take an exclusive ref. Returns true iff it succeeds.
+  inline bool TryExclusiveRef() {
+    uint32_t will_be_deleted = refs & WILL_BE_DELETED;
+    uint32_t expected = will_be_deleted;
+    return refs.compare_exchange_strong(expected,
+                                        EXCLUSIVE_REF | will_be_deleted);
+  }
+
+  // Repeatedly tries to take an exclusive reference, but stops as soon
+  // as an external reference is detected (in this case the wait would
+  // presumably be too long).
+  inline bool TrySpinExclusiveRef() {
+    uint32_t expected = 0;
+    uint32_t will_be_deleted = 0;
+    while (!refs.compare_exchange_strong(expected,
+                                         EXCLUSIVE_REF | will_be_deleted)) {
+      if (expected & EXTERNAL_REFS) {
+        return false;
+      }
+      will_be_deleted = expected & WILL_BE_DELETED;
+      expected = will_be_deleted;
+    }
+    return true;
+  }
+
+  inline void ReleaseExclusiveRef() { refs.fetch_and(~EXCLUSIVE_REF); }
+
+  // The following functions are for upgrading and downgrading refs.
+  // They guarantee atomicity, i.e., no exclusive refs to the handle
+  // can be taken by a different thread during the conversion.
+
+  inline void ExclusiveToInternalRef() {
+    refs += kOneInternalRef;
+    ReleaseExclusiveRef();
+  }
+
+  inline void ExclusiveToExternalRef() {
+    refs += kOneExternalRef;
+    ReleaseExclusiveRef();
+  }
+
+  // TODO(Guido) Do we want to bound the loop and prepare the
+  // algorithms to react to a failure?
+  inline void InternalToExclusiveRef() {
+    uint32_t expected = kOneInternalRef;
+    uint32_t will_be_deleted = 0;
+    while (!refs.compare_exchange_strong(expected,
+                                         EXCLUSIVE_REF | will_be_deleted)) {
+      will_be_deleted = expected & WILL_BE_DELETED;
+      expected = kOneInternalRef | will_be_deleted;
+    }
+  }
+
+  inline void InternalToExternalRef() { refs++; }
```

Comment:
Without referring to the following lines, the "refs++;" is not intuitively understandable. Please add a comment.
static constexpr uint32_t kOneInternalRef = 0x8000;
  static constexpr uint32_t kOneExternalRef = 0x8001;